doc: Write the How do we Test and Document our C API? article (#33)
wasmerio/wasmer.io: _posts/how-do-we-test-and-document-our-c-api.md

---
title: How do we Test and Document our C API?
excerpt: 'Wasmer has designed a unique solution to test and document its own C API.'
date: '2021-07-06T10:32:01.123Z'
author: Ivan Enderlin
published: true
---

> [The Wasmer runtime](https://github.com/wasmerio/wasmer) comes in
> multiple flavors. Written in Rust, it is embedded in different

Reviewer: This will improve SEO for people searching for WebAssembly runtimes in Google.

```suggestion
Wasmer [server-side WebAssembly runtime](https://github.com/wasmerio/wasmer) comes in
multiple flavors. Written in Rust, it is embedded in different
```

([discover all the
integrations](https://github.com/wasmerio/wasmer#-language-integrations)).

As many of our embeddings, the C embedding is implemented in Rust
itself. Rust is perfectly capable of exposing a C compatible ABI
(learn more by reading [The Rustnomicon Book, the Foreign Function
Interface
Chapter](https://doc.rust-lang.org/nomicon/ffi.html)). Moreover, Rust
comes with a first-class tooling for testing and documenting
projects. We wanted to leverage those tools to:

1. Test with `cargo test --lib` and with `cargo test --doc`,
2. Document with `cargo doc`.

Why? Because they are simple, they work everywhere, and every Rust
programmer knows them. They are part of the foundation of the
Rust ecosystem.

## Testing: let's write C directly in Rust

If you are a Rust developer, you are familiar with the `#[test]`
attribute that annotates a function that acts as a test:

```rust
#[test]
fn my_test() {
    let a = 3;
    let b = 1 + 2;
    assert_eq!(a, b);
}
```

To run that test, one simply runs `cargo test --lib`.

If you have ever written tests in C, you know the struggle. `CMake`,
`Makefile`, you name it. What if we could write Rust code to test our

> C API? Something like:

Reviewer: ```suggestion
If you have ever written tests in C you know the struggle. Either with `CMake`,
or `Makefile`, you name it.

But what if we could write Rust code to test our C API?
Something like:
```

> If you have ever written tests in C, you know the struggle. `CMake`,
> `Makefile`, you name it. What if we could write Rust code to test our
> C API? Something like:

Reviewer: ```suggestion
If you have ever written tests in C you know the struggle. Either with `CMake`,
or `Makefile`, you name it.

But what if we could write Rust code to test our C API?
Something like:
```

---

> title: How do we Test and Document our C API?

Reviewer: This title gives very little context to someone that sees it for the first time (who is we, what is our stack in?)

Something more relatable should be:

```suggestion
title: Developing and Testing C libraries from Rust
```
